Assets/Scripts/StandardSamples/UI/Common/SampleMenu.cs

```diff
@@ -0,0 +1,316 @@
+        {
+            IsDirty = isDirty;
+        }
+
+        /// <summary>
+        /// This function is called before refreshing the <see cref="UIFriendsMenu"/>.
+        /// This is an opportunity to refresh any caches or do processing work that should be done once,
+        /// instead of processed for each call of <see cref="GetFriendInteractionState(FriendData)"/>.
+        /// </summary>
+        public virtual void OnFriendStateChanged()
+        {
+
+        }
+    }
+
+    /// <summary>
+    /// Contains implementation common to all sample menus.
+    /// </summary>
+    public abstract class SampleMenu : MonoBehaviour
+    {
+        /// <summary>
+        /// Editable within the Unity Editor, it should be set to the control
+        /// which should have focus when the SampleMenu first becomes visible.
+        /// </summary>
+        [Header("Controller")] 
+        public GameObject UIFirstSelected;
+
+        /// <summary>
+        /// Editable within the Unity Editor, it should be set to the parent
+        /// of the SampleMenu in the Scene hierarchy.
+        /// </summary>
+        public GameObject UIParent;
+
+        /// <summary>
+        /// Indicates whether the SampleMenu is hidden or not. If unset (has no
+        /// value) that indicates that it should be hidden, but has not yet
+        /// been set to be hidden explicitly.
+        /// </summary>
+        private bool? _hidden;
+
+        /// <summary>
+        /// Indicates whether the SampleMenu is hidden or not.
+        /// </summary>
+        public bool Hidden
+        {
+            get
+            {
+                return (!_hidden.HasValue) || _hidden.Value;
+            }
+            private set
+            {
+                _hidden = value;
+            }
+        }
+
+        /// <summary>
+        /// Indicates whether the SampleMenu requires a user be authenticated
+        /// in order to interact with the scene.
+        /// </summary>
+        public bool RequiresAuthentication { get; private set; }
+
+        /// <summary>
+        /// Indicates whether the SampleMenu starts out hidden or visible.
+        /// NOTE: This Property might be something that can be removed, it
+        ///       exists primarily to maintain functionality of current scenes.
+        /// </summary>
+        public bool StartsHidden { get; private set; }
+
+        /// <summary>
+        /// Constructor for SampleMenu.
+        /// </summary>
+        /// <param name="startsHidden">
+        /// Indicates whether the SampleMenu should be visible or hidden when it
+        /// launches. Most of them should be hidden.
+        /// </param>
+        /// <param name="requiresAuthentication">
+        /// Indicates whether the SampleMenu requires a user be authenticated in
+        /// order for it to be used. Most of them require authentication.
+        /// </param>
+        protected SampleMenu(bool startsHidden = true, bool requiresAuthentication = true)
```

Comment:
MonoBehaviours cannot have constructors. I think?

Reply:
While this is true (and something I just learned) the reason that this _does_ work is because it sets what are effectively readonly parameters. Unity will not honor or support MonoBehaviour implementing classes having a constructor with parameters, since `SampleMenu` is abstract, and implementing classes all set the variables themselves, the result is that the type constructed by Unity will be parameterless, and therefore _will_ be executed.